src/main/java/ch/njol/skript/lang/parser/ParsingStack.java

```diff
@@ -0,0 +1,222 @@
+/**
+ *   This file is part of Skript.
+ *
+ *  Skript is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package ch.njol.skript.lang.parser;
+
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.SkriptParser;
+import ch.njol.skript.lang.SyntaxElement;
+import ch.njol.skript.lang.SyntaxElementInfo;
+import ch.njol.util.Kleenean;
+
+import java.io.PrintStream;
+import java.util.Iterator;
+import java.util.LinkedList;
+
+/**
+ * A stack that keeps track of what Skript is currently parsing.
+ * <p>
+ * When accessing the stack from within {@link SyntaxElement#init(Expression[], int, Kleenean, SkriptParser.ParseResult)},
+ * the stack element corresponding to that {@link SyntaxElement} is <b>not</b>
+ * on the parsing stack.
+ */
+public class ParsingStack implements Iterable<ParsingStack.Element> {
```

Comment:
This seems like it's duplicating a lot of code from classes like [Deque](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/util/Deque.html)... Could we extend that instead?

Reply:
I didn't want to extend any of the main stack implementation as it'd allow operations that aren't appropriate for the parsing stack. For example modifying both the start and end, or removing specific elements from halfway throughout the deque. Having fewer methods also makes the API simpler IMO

Reply:
Sounds good to me :)